python/lsst/cp/pipe/utilsEfd.py

```diff
@@ -0,0 +1,151 @@
+    ----------
+    efdInstance : `str`, optional
+        EFD instance name to connect to.
+    log : `logging.Logger`, optional
+        Log to write messages to.
+    """
+
+    def __init__(self, efdInstance="usdf_efd", log=None):
+        self.log = log if log else logging.getLogger(__name__)
+
+        nest_asyncio.apply()
+        if haveEfd:
+            self.client = EfdClient(efdInstance)
+        else:
+            self.client = None
+            self._emitWarning()
+
+    def _emitWarning(self):
+        if not haveEfd:
+            self.log.warning("EFD client not available.")
+            return None
+
+    def close(self):
+        """Delete self, which will close all open connections."""
+        del self
+
+    def getEfdMonochromatorData(self, dataSeries=None, dateMin=None, dateMax=None):
+        """Retrieve Electrometer data from the EFD.
+
+        Parameters
+        ----------
+        dataSeries : `str`, optional
+            Data series to request from the EFD.
+        dateMin : `str`, optional
+            Minimum date to retrieve from EFD.
+        dateMax : `str`, optional
+            Maximum date to retrieve from EFD.
+
+        Returns
+        -------
+        results : `pandas.DataFrame`
+            The table of results returned from the EFD.
+        """
+        self._emitWarning()
+        # This is currently the only monochromator available.
+        dataSeries = dataSeries if dataSeries else "lsst.sal.ATMonochromator.logevent_wavelength"
+
+        if dateMin:
+            start = Time(dateMin, format='isot', scale='utc')
+        else:
+            start = Time("1970-01-01T00:00:00", format='isot', scale='utc')
+        if dateMax:
+            stop = Time(dateMax, format='isot', scale='utc')
+        else:
+            stop = Time("2199-01-01T00:00:00", format='isot', scale='utc')
+
+        loop = asyncio.get_event_loop()
+        results = loop.run_until_complete(self.client.select_time_series(dataSeries, ['wavelength'],
+                                                                         start.utc, stop.utc))
+        results.sort_index(inplace=True)
+        return results
+
+    def parseMonochromatorStatus(self, data, dateStr):
+        """Determine monochromator status for a specific date.
+
+        Parameters
+        ----------
+        data : `pandas.DataFrame`
+            The dataframe of monochromator results from the EFD.
+        dateStr : `str`
+            The date to look up in the status for.
+
+        Returns
+        -------
+        indexDate : `str`
+            Date string indicating the monochromator state change.
+        wavelength : `float`
+            Monochromator commanded peak.
+        """
+        dateValue = Time(dateStr, format='isot', scale='utc')
+        # I have words here.
+        if data.index[0] > dateValue or data.index[-1] < dateValue:
+            raise RuntimeError("Requested date is outside of data range.")
+
+        # I have more words.
```

Comment:
???

Reply:
"Binary search through the EFD entries in date, until the most recent monochromator state update prior to the spectrum in question is found."
If there's a better way to search through a pandas dataframe, I'm happy to use that, but I was unable to find one with some searching.